### ISSUE TITLE
Good disjoint unions, algebraic datatypes, and so on

### ISSUE THREAD

**Issue by jeapostrophe** [racket/racket#3533]
Disjoint unions and algebraic data-types are useful. For example, a binary tree is either a leaf or a branch. I've heard binary trees are really useful.

In Racket, we might implement them like this:
```
(struct leaf (val))
(struct branch (val left right))

;; A binary tree is either
;; - a (leaf value); or,
;; - a (branch value binary-tree binary-tree)
```

This is a bit unsatisfying because the comment is not a program. There are a bunch of different ways of trying to make that comment program-like.

For example, we could make a predicate function:
```
(define (binary-tree? x)
  (or (leaf? x) (branch? x)))
```

This is kind of unsatisfying though, because it doesn't fully check the structure or integrate well with other pieces of Racket. For instance, we might want to use the contract system instead:
```
(define binary-tree/c
  (flat-rec-contract bt
    (or/c leaf? (struct/c branch [value any/c] [left bt] [right bt]))))
```
This would then give us access to `contract-random-generate` and so on.

There are other parts of Racket that we might like to integrate with, such as `match`. The predicate function style allows us to write
```
(match x
  [(? binary-tree?) ....])
```
but the contract style doesn't. And neither allows us to ensure that a `match` expression handles all variants of binary trees:
```
;; Good
(match x
  [(leaf v) ....]
  [(branch v l r) ....])

;; Bad
(match x
  [(leaf v) ....])
```

Some way of having all these things is a popular idea in programming languages and Racket doesn't really have _a way_ of doing this. Instead, it gives a bunch of tools that can be composed to make something similar. Many people have attempted to add something like this to Racket. For example, `2htdp/abstraction`, `plai`, `plai-typed`, `plait`, and `plaitypus` define `define-type`, while `lang/htdp-advanced`, `eopl`, and `typed-racket-datatype` define `define-datatype`, and `algebraic/data` defines `data`. Each of these is Not Quite Right, in that they don't fully integrate with all parts of Racket and some have the parentheses in the wrong places.

I think that there should be `racket/data` (or something like that) that integrates well with all of the existing Racket infrastructure, like `struct`, `racket/generic`, `racket/contract`, `racket/match`, and so on.

I'd like to use this issue to talk about the design of the feature and then go off and build it.

My non-goal is to make something that would be used instead of `plai` in a PLAI-based class. My goal is to recognize that this is a generally useful programming construct that Racket doesn't adequately support. If this new generally useful feature means that some existing clients of similar libraries don't want to exist anymore, that's cool.

Here are some issues to discuss:

Scope:
- [ ] Decide that datatypes are fundamentally static things; or,
- [ ] Decide that they are dynamic and can be extended at runtime; or,
- [ ] Decide that they are static, but they can be extended at compile-time, meaning that the definition of a datatype can be in different modules (i.e. outside of one set of parens) [think type-classes in Haskell, including orphan and overlapping issues]

Definition:
- [ ] Extension to `struct` that says "I am part of data-type X"; or,
- [ ] New `define-disjoint-union-thingy` form that says "These structs are variants of me"; or,
- [ ] New 'define-datatype' form that defines the data-type and the variants and supports all the features of `struct`

Exhaustiveness:
- [ ] Extension to `racket/match` that does static checking that cases are exhaustive for some general notion of exhaustive; or,
- [ ] New `match-with-data-that-I-know-how-to-ensure-exhaustive` macro?

Contracts:
- [ ] Extension to `struct` that builds in contract-checking (like `plai` does); or,
- [ ] Don't do that, but maybe generate a `datatype/c` binding on the side

Implementation:
- [ ] Where does this implementation live in the language hierarchy? Is it allowed to use `syntax/parse`? Does it have to be written in `#%kernel`? Does it affect the VM in any way, like with a special kind of a chaperone?

**Comment by jeapostrophe** [comment]
I think that these people might have important useful things to add about this topic @mflatt @mfelleisen @rfindler @samth @AlexKnauth @dedbox @pavpanchekha

**Comment by rfindler** [comment]
Enumeration support too! Should there be a connection to TR?

I feel like a lot of the issues raised here about the design are ones where the PL community know what a good design would look like, but the extensibility part is one where there's a lot less consensus. That's not to say we shouldn't do that, but perhaps the best thing would be to either leave that out or leave some way for that to be added "from the outside" so we don't have to commit to that part of the design very early.

Another point that seems like worth making explicit is that some of the things here require more information than others. So a design that accommodates different amounts of information about the data type that's being defined in order to support different things might make sense. And do we want this to be open in the sense that, say, the data/enumerate library could say how to add enumeration to one of these without a dependency from this library to data/enumerate but instead the other way?

**Comment by dedbox** [comment]
Good timing. I've been thinking about Algebraic Racket's data API for a week now. Here's some of what I think would enhance extensibility while maintaining a clean, minimal phase-1 footprint.

Constructors should be defined individually instead of in named groups. For example, `Nothing` and `Just` constructors should be definable independently from any `Maybe` type. It's really the type system's job to determine (statically) how constructors are created, compared, and ordered. On the other hand, primitive type-specific functionality like equality checks and reading/printing, or common behaviors implemented as struct properties or generic methods, should have safe defaults like `equal?` and `read`/`write` that can be overridden, so constructors can still be useful without having to bolt on a type system.

Each constructor should be able to have a different underlying run-time implementation. The default implementation should be something sane, like structs or tagged lists, that can be replaced with something more esoteric like hashes, rebellion record types, or generator functions. Implementations should be defined in a manner similar to classes or traits, so they can be mixed or extended easily at run time.

With this arrangement, implementers *and* users are free to mix and match features to suit their needs. Pluggable types are a good fit for language-oriented programming and ADTs are a good case study for them, IMO. @david-christiansen 's TyDe 2020 talk on Klister has some interesting ideas on pluggable types in Racket.

Some type-directed functionality I'm targeting:

- Exhaustivity checks over ad hoc type unions
- Replacing dynamically dispatched generics with statically generated, type-specific implementations
- Cheap analysis / instrumentation of run time executions (e.g., information flow control, abstract interpretation)

**Comment by jackfirth** [comment]
Note also that this is an option today:

```scheme
(struct binary-tree ())
(struct leaf binary-tree (val))
(struct branch binary-tree (val left right))
```

**Comment by jeapostrophe** [comment]
Thanks Jack, I meant to include that but forgot :)

**Comment by jackfirth** [comment]
Since structs already have subtyping, I think we should avoid making some new separate system that creates an orthogonal axis from the subtyping hierarchy.

There's a neat solution to this problem in Java called "sealed classes", which lets you mark classes/interfaces as only allowing a fixed set of inheritors, known at compile time. For example, the binary tree ADT example in Java using sealed types and records (a Java 14 feature, they're basically immutable transparent structs) would look something like this:

```java
sealed interface BinaryTree<T> permits Leaf<T>, Branch<T> {}

record Leaf<T>(T val) extends BinaryTree<T> {}

record Branch<T>(T val, BinaryTree<T> left, BinaryTree<T> right) extends BinaryTree<T> {}
```

And then you can write a switch statement on a `BinaryTree` that has exhaustiveness checking.

**Comment by samth** [comment]
My preferences are
- just a new `define-datatype` form that expands to the pattern @jackfirth describes
- no special VM or contract system or struct system support
- but with the full kinds of customization that the struct system supports (such as arbitrary properties/inspectors/etc).

This seems like the thing that people are regularly inventing and re-purposing from elsewhere. 

For exhaustiveness checking, I imagine that we could add the following to `match`:
- a new kind of struct pattern with a pointer to a list of patterns
- a check that each one of those was matched against
That wouldn't be full exhuastiveness checking, since nested patterns might fail to match causing the whole thing to fail. But neither disallowing nested matching nor trying to do exhaustiveness checking on general match patterns seems like a good idea.

**Comment by jackfirth** [comment]
@samth Is there a way to prevent users of `define-datatype` from creating new cases that inherit from the parent struct type? I'm not sure I want a form like this if it can't enforce that unexpected subtypes aren't created.

**Comment by samth** [comment]
@jackfirth Just not exposing the structure type would be sufficient for that.

**Comment by rfindler** [comment]
It would be nice to be able to avoid the extra constructors that show up in `define-type` (and in @samth's suggestion) somehow. When you already have a predicate for something, just being able to use it seems helpful.

**Comment by pavpanchekha** [comment]
I've needed something like this lots of times, but let's try to separate out the possible goals:

- A convenient syntax for tagged unions
- A consensus runtime representation for tagged unions
- Nicer error messages for runtime type errors
- Nicer error messages for failing to match all variants of tagged unions
- A guarantee that certain tagged unions won't be extended
- More common use of tagged unions throughout the Racket libraries

Structs with subtyping are a classic representation (@jackfirth) but are inconvenient syntax-wise (because there's no guarantee all subtypes are in one place), give bad error messages, and don't have exhaustiveness checking. They integrate nicely with some Racket features; obviously type classes are nice but Racket uses generic functions instead and structs allow that. Structs carry costs: structs have complex options our ADT system would need to consider (mutability? prefab?) and they have long ugly method names that I at least do not like. Plus they print funny. Some of these can be fixed with syntax (we could generate a nicer string representation) but that makes it all more complex (what if I want custom printing?) and brittle. And note that even adding an ADT library won't mean the racket core uses it. `dict-ref` will still have a `#:default` argument instead of returning a `Maybe`, you know?

Look, Racket isn't Haskell, it's Lisp, and it inherits all sorts of design decisions from it. Maybe it's better this way? We could instead make it easier for users to define-their-own ADT-like things. For example in Cassius I use the following macro, which lets me turn :

```
(define-syntax-rule (define-match/c name patterns ...)
  (define name
    (flat-contract
      'name
      (λ (var)
       (let name ([var var])
         (match var
           [patterns true] ...
           [_ false]))))))
```

That lets me add contracts for poor-man's-ADTs, like this (for CSS selectors):

```
(define-match/c selector?
  `*
  `(id ,(? symbol?))
  `(class ,(? symbol?))
  `(tag ,(? symbol?))
  `(and ,(? selector?) ...)
  `(desc ,(? selector?) ...)
  `(child ,(? selector?) ...)
  `(pseudo-class ,(or 'first-child 'last-child 'hover 'last-of-type 'first-of-type))
  `(type ,(? symbol?))
  `(media ,(? media-query?) ,(? selector?))
  `(fake ,(? string?) ,(? selector?) ...)))
```

Racket's `match` is quite powerful so using it for contracts is great. And because the `define-match/c` macro embeds the `define`, it can be used for recursive data structures much like an ADT. That said, error reporting with this macro is not great: it'll tell you about the top-level match failure but usually there's a more specific cause that it won't help you pinpoint. Maybe that can be fixed.

Anyway, my contribution to the discussion is to say: embedding ADTs deep into Racket would make it a new language. Maybe Rhombus should have them, but adding an ADT library to Racket isn't going to dramatically change the rest of the language. To me that makes concerns like exhaustiveness checking and sealed types difficult to argue for—nothing else in Racket works like that. The problems that Racket does claim to solve—like nice yet flexible syntax or a powerful contract system—should be the focus.

**Comment by mfelleisen** [comment]
Thank you Jay for laying out the design space for disjoint union and friends. Your description is quite operational, so let me throw in a denotational view. 

Racket has walked a fine line between being too dynamic and too static. You weren't around when we were accused of betraying Scheme because we injected "static structs" kind of like ML. (I completely fail to understand the comment about Racket inheriting anything from Haskell.)  But structs were carefully evolved _not_ to be ML, which has only one extensible data type (exe). So one 'denotational' question is the dynamic-static one. Finding _our_ exact spot is what will distinguish Racket from the rest. 

Racket has also been quite careful of what goes into the core. As someone else said already, the idea of such (possibly disjoint) unions should not be "down there". But where should they be located? We don't know what we want for sure --- otherwise you wouldn't' have opened an issue -- so I think you need at least two phases: 

1. a formative language design  phase that leverages all of Racket to get the "thing" that we all embrace as the correct addition to Racket. 

2. a commitment and an investigation whether this should be "lowered" in the language hierarchy. 

We will always be "stuck" in phase 1, and that may be okay. 

Finally, Racket is fundamentally a dynamically typed language not a statically typed one. This means "we" start from a "universal data space" and narrow it down to the part we want, even if we make some constructions somewhat static. The most general definition mechanism would be a projection-based one. No, we don't need this generality and extreme generality rarely serves programmers. But we could at least settle for a set-based approach -- specify the union we want after the sets are established, and perhaps abstract over this. In case this isn't clear, I just want to say that "disjoint:" is fundamentally flawed; "tidy" might be okay.  

[[ I also think that the discussion is seriously under-informed when it comes to `datatype`. ML very clearly demonstrates that these are syntactic types -- in the sense of Reynolds -- and not what we call, loosely and in a losing fashion, "types".  The distinction between `datatype` in a module signature and in a module implementation brings this directly to mind. In this sense, my desire for a plain union isn't actually all that far from ML -- it's just "more dynamic". ]]

@jeapostrophe

**Comment by jeapostrophe** [comment]
Matthias made this other comment to me that I think should go here:

"I think for a dynamically typed language the correct construct is a set-oriented approach to incrementally defining sets of data. The ‘incremental’ part here means across modules. For computability and efficiency, we may wish to impose some constraints."

and

"The other aspect is that such definitions should come with a way to connect to `match`.  I imagine that such definitions can add something that `match` can "pick up” statically to check some aspects (perhaps not exhaustiveness, but coverage “up to”) of the definition."

To understand this in my original rubric, I'm reading this as saying that something like `define-set` is wanted and something like `match-in-set`, although perhaps with some way of having normal `match` "pick up" on it. In other words, that the thing being emphasized is that we are adding a *new* thing, rather modifying *old* things (like `struct`s). My main question is what `define-set` is useful for at runtime... is it just a predicate that checks if a value is in it? It sounds like a contract, so it would be feasible to connect in the enumerations and what not.

I agree with Matthias that this "feels" very dynamic-typey, but I also suspect that it is not the only thing that people want, because they really do want to write ML in Racket. I feel like there's tension on this issue between these two points: do we make a canonical way to "do ML in Racket" and/or do we try to make "the Racket way" of addressing the same underlying motivation?

**Comment by mfelleisen** [comment]
Correct. As much as I have pushed Racket into the direction of more statics because of my affinity to ML in the 90s, I also always kept in mind that if I want to program in ML (or Haskell), I can program in some implementation of ML. What Racket brings to the table is a more flexible way of reasoning about data, of gluing together components, and distributing definitions across components ~~~ and we haven't even mentioned macrology yet. I find this way of programming more fascinating than ML and friends, and speaking as a researcher, I'd like to explore this direction (and if possible convince some people that it is a cool direction).

**Comment by samth** [comment]
Further research on how to do more flexible ways of managing data is great, but currently we see lots of people wanting something basically like `define-datatype` in `#lang racket`, and that's more straightforward to address without new research.

**Comment by dedbox** [comment]
I agree emphatically with both Matthias and Sam, so I must say that doing research in this direction is not necessarily the same as making this sort of research possible.

In the spirit of the latter, my current draft of Algebraic Racket's new data API has two fundamental forms: `define-data-implementation` and `define-data-constructor`.

`define-data-implementation` defines a transformer binding that specifies how to construct and match against an existing Racket data type, and how to define type-specific helpers. As a language / library designer, this gives me the freedom to choose an implementation that suits my needs. For now, I've got unreadable symbols for nullary constructors (i.e., constants) and lists for everything else, in addition to a variety of more esoteric alternatives. This arrangement allows constructors to inherit the unmitigated functionality of their underlying types, so ancillary functionality like equality checks and ordering do not need to touch the API.

`define-data-constructor` defines a transformer binding that determines whether its identifier is being used as a constructor function or a match expander and then expands to another identifier bound to something implementation specific, as a sort of expansion-time dispatch. This works well in practice, with the caveat that matching constants is a little clunky because `match` does not support identifier macros.

I am deliberately omitting a `define-data-type` form at this point in the API because I am still working toward a sufficiently language-oriented typing framework. Somewhere in here, I'm guessing, is where the potential for research begins, but there are useful things to do in this direction before wandering into the weeds. A `define-closed-union` form that expands to a batch of `define-data-constructor` calls and cooperates with a `match`-like form for exhaustivity checks and nicer error messages is certainly feasible, not novel, and in the spirit Matthias advocates.

As a side note, I'm only using Algebraic Racket to exemplify the sort of flexibility I look for from Racket as a platform for language-oriented programming, and to demonstrate what I know is possible. At the end of the day, I'm all for whatever gives us the most flexibility for the least cost.
